bottlecap/src/metrics/dogstatsd.rs

```diff
@@ -35,63 +56,115 @@ impl DogStatsD {
+        }
+    }
+
+    async fn consume_statsd(&self) {
+        let (buf, src) = self
+            .buffer_reader
+            .read()
+            .await
+            .expect("didn't receive data");
+        let msgs = std::str::from_utf8(&buf).expect("couldn't parse as string");
+        debug!(
+            "received message: {} from {}, sending it to the bus",
+            msgs, src
+        );
+        let statsd_metric_strings = msgs.split('\n');
+        self.insert_metrics(statsd_metric_strings);
+    }
+
+    fn insert_metrics(&self, msg: Split<char>) {
+        let all_valid_metrics: Vec<Metric> = msg
+            .filter_map(|m| match Metric::parse(m) {
+                Ok(metric) => Some(metric),
                 Err(e) => {
-                    error!("failed to parse metric: {:?}\n message: {:?}", msg, e);
-                    continue;
+                    error!("failed to parse metric: {}", e);
+                    None
                 }
-            };
-            if parsed_metric.name == "aws.lambda.enhanced.invocations" {
-                debug!("dropping invocation metric from layer, as it's set by agent");
-                continue;
+            })
+            .collect();
+        if !all_valid_metrics.is_empty() {
+            let mut guarded_aggregator = self.aggregator.lock().expect("lock poisoned");
+            for a_valid_value in all_valid_metrics {
+                let _ = guarded_aggregator.insert(&a_valid_value);
             }
-            let first_value = match parsed_metric.first_value() {
-                Ok(val) => val,
-                Err(e) => {
-                    error!("failed to parse metric: {:?}\n message: {:?}", msg, e);
-                    continue;
-                }
-            };
-            let metric_event = MetricEvent::new(
-                parsed_metric.name.to_string(),
-                first_value,
-                parsed_metric.tags(),
-            );
-            let _ = self
-                .aggregator
-                .lock()
-                .expect("lock poisoned")
-                .insert(&parsed_metric);
-            // Don't publish until after validation and adding metric_event to buff
-            let _ = self.event_bus.send(Event::Metric(metric_event)).await; // todo check the result
-            if self.cancel_token.is_cancelled() {
-                debug!("closing dogstatsd listener");
-                break;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::config::Config;
+    use crate::metrics::aggregator::Aggregator;
+    use crate::metrics::aggregator::ValueVariant::Value;
+    use crate::metrics::dogstatsd::{BufferReader, DogStatsD};
+    use crate::tags::provider::Provider;
+    use crate::LAMBDA_RUNTIME_SLUG;
+    use std::collections::HashMap;
+    use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+    use std::sync::{Arc, Mutex};
+
+    #[tokio::test]
+    async fn test_dogstatsd() {
```

Comment:
Would it make sense to also add a single metric test? here we are testing multiple metrics at once, WDYT?

Reply:
I think the string parsing should be tested as unit test in the metric parser.
I will add a 1 metric example just to make sure the newlines won't play fun